### DISCUSSION TITLE
How to add phone number and require verification on existing user?

### DISCUSSION THREAD

**Discussion by chekdotdev** [supabase/supabase#7265] [Questions] [answered]
We are creating accounts using Email/OAuth (Google) and want to utilize `phone` inside the `auth.users` table for its built-in Twilio and verification support.

However, we're unable to update phone numbers **and** require phone number verification. After adding the phone number to the user via `PUT /auth/v1/users`, Supabase returns a `404` when trying to submit the OTP via `/auth/v1/verify`:

<img width="1208" alt="Screen Shot 2022-06-13 at 11 08 20 AM" src="https://user-images.githubusercontent.com/28504852/173386229-4f0e8707-f49f-443f-bc3c-8b2e1a8cb034.png">
 
<img width="1210" alt="Screen Shot 2022-06-13 at 11 10 57 AM" src="https://user-images.githubusercontent.com/28504852/173386241-6ccb5422-88a0-4554-ac96-ea259a365b11.png">

It appears that the `new_phone` value is there, but Supabase does not find the user. Is there another route we should be using to add and *separately* verify a phone number on an existing user? 

We also cannot verify at `/auth/v1/verify` using the verified email address that is tied to the user:

<img width="458" alt="Screen Shot 2022-06-13 at 11 15 40 AM" src="https://user-images.githubusercontent.com/28504852/173386633-b0301b79-960f-4adb-82f5-8341ab633840.png">

**Comment by GaryAustin1** [comment, marked as answer]
I've not used phone numbers for auth.  But looking at gotrue code for verify, in order to check the "new" phone number, which gets stored in phone_change in the auth.user table, you need to provide "phone_change" instead of "sms" in the type field.  With "sms" it looks at the original phone number field which would be "" so won't match.

**Reply by chekdotdev** [reply]
This worked, thank you @GaryAustin1!

We are using the [supabase-js](https://github.com/supabase/supabase-js) client, so the function looks like this:

```js
let { session, error } = await supabase.auth.verifyOTP({
    phone: `+1${phone}`,
    token: otp,
    type: "phone_change"
})
```

Putting the gotrue-js reference here for anyone who needs it:
https://github.com/supabase/gotrue-js/blob/30fb5729a1988dbd7f61d58ec10eb81de63f93c7/src/GoTrueApi.ts#L373-L401

**Reply by growthondev** [reply]
does dart library has "type" parameter?

**Comment by no2pixel** [comment]
I'm trying to implement this same type of verification flow (using supabase.js, not dart), and everything seems to work/update properly when adding the phone number, but when trying to login later using the updated phone number, either with password or with SMS OTP, I receive a "User not found" or "Invalid Credentials" error.

Is it even possible to have one account be able to log in using email/password and phone/otp?

**Reply by ColeTownsend** [reply]
@no2pixel when you are sending that request, you need to first do a `PUT` to initiate the phone change. Then you must send a verification token like so
```
let { session, error } = await supabase.auth.verifyOTP({
    phone: `+1${phone}`,
    token: otp,
    type: "phone_change"
})
```

or http
```
curl --request POST \
  --url {{SUPABASE URL}}/auth/v1/verify \
  --header 'Content-Type: application/json' \
  --header 'apikey: {{API KEY}}' \
  --data '{
	"type": "phone_change",
  "phone": "{{YOUR PHONE HERE}}",
	"token": "{{OTP RECEIVED}}"
}'
```

**Comment by harshalone** [comment]
Hi I am trying to update user phone number after authentication.

Hi I am trying to update user phone number after authentication.

to do so I am using this function with supabase-js v2

```
const {data: { session },} = await supabase.auth.updateUser({
        phone: ${mobile},
    });
```

it sends user the OTP but I can see in the auth users table the phone number is not updated. Also when I verify the users mobile number with OTP it does not work.

ps: using message bird for OTP message.

Do you know what I am doing wrong here?

**Reply by sioquim** [reply]
@harshalone did you manage to get this sorted? I'm having a similar issue.

**Reply by ColeTownsend** [reply]
You must then follow a similar flow to signing up. After you get OTP you must do:
```
let { session, error } = await supabase.auth.verifyOTP({
    phone: `+1${phone}`,
    token: otp,
    type: "phone_change"
})
```

**Comment by ChinKX** [comment]
Hi! I just want to share a more detailed step-by-step guidelines on how I link phone number to an existing oauth user.

I actually implemented google sign-in for my web app and also ask for user mobile number afterwards. The following Typescript code snippet demonstrates how I link the phone number to an existing oauth user:

```js
const supabaseClient = Supabase.createClient(
  SUPABASE_URL,
  SUPABASE_API_KEY
)

// 1. Update user phone number (Note: Remember to check for error here as supabase might throw error saying that the mobile number already belongs to some other supabase user)
const { error } = await supabaseClient.auth.updateUser({ phone: mobileNumber })
if (error) {
  throw error
}

// 2. Request phone change OTP
const {error} = await supabaseClient.auth.resend({ phone: mobileNumber, type: 'phone_change' })
if (error) {
  throw error
}

// 3. Verify OTP
const {error} = await supabaseClient.auth.verifyOtp({ phone: mobileNumber, token: otp, type: 'phone_change' })
if (error) {
  throw error
}
```

Feedbacks/questions are welcome! Hope this helps and thanks!

**Reply by hugoh59** [reply]
Following exactly what you did, I'm still getting this error: 

`Error verifying OTP: [AuthApiError: Token has expired or is invalid]`
